### ISSUE TITLE
Add autocompletion support for custom languages

### ISSUE THREAD

**Issue by mdboom** [iodide-project/iodide#804]
We should have a hook so that languages other than Javascript can perform autocompletion.

**Comment by louisabraham** [comment]
I think it would be awesome to implement an interface for https://langserver.org/, at least for code completion and jump to def.

This is a standard that aims to become a standard to implement custom language features for editors.
It is currently supported by the major editors and languages.
Thus, one would just have to compile the language server to JS / WASM to have a full-featured editor in the browser.

**Comment by bcolloran** [comment]
this is the first i've heard of langserver, and i think it sounds like a really interesting idea.

@louisabraham would you be interested in looking into this a bit as you explore ocaml integration? We'll put this on the roadmap, but we probably won't get to if for a bit, so if it's an itch you're excited to scratch we'd love to help you work through a PR! Thanks!

**Comment by mdboom** [comment]
I agree that using langserver as a base means we don't have to reinvent that API or any of the language backends -- just a little glue to handle the events in iodide.  I agree it might be a while until we get to this, but it sounds like a good plan.  Keep us posted on progress on this, @louisabraham, and I'll do the same when I find the time to do the Python autocompletion -- it would be great, as you suggest, to do all of that in a language-agnostic way.

**Comment by louisabraham** [comment]
Disclaimer: I discovered langservers a month ago. But I like it when I'm the first one to have the idea of connecting two cool open source projects.

I have to admit I am not an an expert of neither OCaml nor JS, but I'll definitely work on something usable for OCaml. If you wonder where my obsession with OCaml comes from, it is because I once [dreamt](https://github.com/andrewray/iocamljs/pull/12) of a usable OCaml-in-browser experiment, mostly to help fellow students getting started with OCaml.
The best would be a [real toplevel compiled with js_of_ocaml](https://try.ocamlpro.com/), but an alternative would be to use [BuckleScript](https://bucklescript.github.io/). End of the OCaml off-topic :)

**I don't think one should mix the difficulties and try to implement autocompletion for Python or OCaml first.**

For Python it will probably require including a lot of dependencies, and the resulting experiment might not be very performant / responsive. A good idea would be to execute those with a faster python-in-browser interpreter like http://pypyjs.org/ or to transpile them with https://github.com/flexxui/pscript (the most performant IMO).
So it might get very complicated, and become a big standalone project like the compilation of numpy&co for pyodide.

To begin, there is [a cool Javascript-Typescript langserver](https://github.com/sourcegraph/javascript-typescript-langserver) that could be integrated by either plugging the JSON-RPC layer or, preferably, wiring the TypeScript interfaces.

Thanks to the standard, adding new langservers will be **free** if they are already written in JS/TS (like [ocaml](https://github.com/freebroccolo/ocaml-language-server))
Else, if will require:
- compiling / transpiling the langserver (see above for the case of Python)
- wiring the API calls

**Comment by bcolloran** [comment]
related to https://github.com/iodide-project/iodide/issues/2141
